src/include/txn_inline.h

```diff
@@ -310,26 +311,64 @@ __wt_txn_op_delete_commit_apply_timestamps(WT_SESSION_IMPL *session, WT_REF *ref
      */
     if (previous_state == WT_REF_DELETED) {
         if (ref->ft_info.del->timestamp == WT_TS_NONE) {
-            ref->ft_info.del->timestamp = txn->commit_timestamp;
-            ref->ft_info.del->durable_timestamp = txn->durable_timestamp;
+            if (in_commit_phase) {
+                /*
+                 * Updates without a timestamp can be retroactively timestamped on transaction
+                 * commit, but if this happens for a partially timestamped transaction then using
+                 * commit_timestamp will lead to out-of-order timestamps. Using
+                 * first_commit_timestamp will guarantee a correct timestamp ordering.
+                 */
+                ref->ft_info.del->timestamp = txn->first_commit_timestamp;
+                ref->ft_info.del->durable_timestamp = txn->first_commit_timestamp;
+            } else {
+                ref->ft_info.del->timestamp = txn->commit_timestamp;
+                ref->ft_info.del->durable_timestamp = txn->durable_timestamp;
+            }
         }
     } else if ((updp = ref->ft_info.update) != NULL)
-        for (; *updp != NULL; ++updp) {
-            (*updp)->start_ts = txn->commit_timestamp;
-            (*updp)->durable_ts = txn->durable_timestamp;
-        }
+        for (; *updp != NULL; ++updp)
+            if ((*updp)->start_ts == WT_TS_NONE)
+                __wt_txn_untimestamped_upd_set_timestamp(session, *updp, in_commit_phase);
 
     WT_REF_UNLOCK(ref, previous_state);
 }
 
+/*
+ * __wt_txn_untimestamped_upd_set_timestamp --
+ *     Set the timestamp for an untimestamped update.
+ */
+static inline void
+__wt_txn_untimestamped_upd_set_timestamp(
+  WT_SESSION_IMPL *session, WT_UPDATE *upd, bool in_commit_phase)
+{
+    WT_TXN *txn;
+    txn = session->txn;
+
+    WT_ASSERT(session, upd->start_ts == WT_TS_NONE);
+
+    if (in_commit_phase) {
+        /*
+         * Updates without a timestamp can be retroactively timestamped on transaction commit, but
+         * if this happens for a partially timestamped transaction then using commit_timestamp will
+         * lead to out-of-order timestamps. Using first_commit_timestamp will guarantee a correct
+         * timestamp ordering.
```

Comment:
Can we also include in the comments why it is okay to set the durable to the first_commit_timestamp? Essentially it'll describe that we can't be in the context of a prepared transaction and in this code path, we should also assert that we are not in a prepared transaction.

Reply:
Added in 53eac3c